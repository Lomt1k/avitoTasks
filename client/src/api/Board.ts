import { z } from "zod";
import { api, validateResponse } from "./api";

export const BoardSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string(),
  taskCount: z.number(),
});

export const BoardResponseSchema = z.object({
  data: z.array(BoardSchema),
});

export type Board = z.infer<typeof BoardSchema>;

export const fetchBoards = async (): Promise<Board[]> => {
  const response = await api.get('/boards');
  validateResponse(response);
  const result = BoardResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}