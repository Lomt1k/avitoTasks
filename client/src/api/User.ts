import { z } from "zod";

export const UserSchema = z.object({
  avatarUrl: z.string(),
  description: z.string(),
  email: z.string(),
  fullName: z.string(),
  id: z.number(),
  tasksCount: z.number(),
  teamId: z.number(),
  teamName: z.string(),
});

export const AssigneeSchema = UserSchema.pick({
  avatarUrl: true,
  email: true,
  fullName: true,
  id: true,
});

export type User = z.infer<typeof UserSchema>;

export type Assignee = z.infer<typeof AssigneeSchema>;