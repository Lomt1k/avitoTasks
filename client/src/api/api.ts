import axios, { AxiosResponse } from "axios";
import { z } from "zod";

export const api = axios.create({
  baseURL: 'http://localhost:8080/api/v1',
  timeout: 1000,
  validateStatus: () => true,
});

export const ErrorMessageSchema = z.object({
  error: z.string(),
  message: z.string(),
});

export type ErrorMessage = z.infer<typeof ErrorMessageSchema>;

export const validateResponse = (response: AxiosResponse): void => {
  if (response.status === 200) {
    return;
  }

  const result = ErrorMessageSchema.safeParse(response.data);
  if (result.data) {
    const error = new Error(result.data.message);
    error.name = result.data.error;
    console.error(error);
    throw error;
  }

  const error = new Error(`${response.status} ${response.statusText}`);
  error.name = 'Network Error';
  console.error(error);
  throw error;
}