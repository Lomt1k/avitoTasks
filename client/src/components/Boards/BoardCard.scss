.board-card {
  padding: 20px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: 20px;
  background-color: var(--color-white);
  border-radius: 15px;
  position: relative;

  @include vp-767 {
    display: flex;
    flex-direction: column;
    align-items: flex-end;
  }

  &__info {
    display: flex;
    flex-direction: column;
    gap: 10px;

    @include vp-767 {
      width: 100%;
    }
  }

  &__info-top {
    display: flex;
    align-items: center;
    gap: 10px;
  }

  &__title {
    font-size: 20px;
    font-weight: 600;
  }

  &__task-counter {
    color: var(--color-white);
    background-color: var(--color-green);
    border-radius: 30px;
    padding: 2px 10px 0;
    font-weight: 500;
    user-select: none;

    @media (max-width: 500px) {
      position: absolute;
      left: 20px;
      bottom: 25px;
      padding: 7px 12px 5px;
    }
  }
}