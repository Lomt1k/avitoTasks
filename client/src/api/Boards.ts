import { z } from "zod";
import { api, validateResponse } from "./api";

export const GeneralBoardInfoSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string(),
  taskCount: z.number(),
});

export const GeneralBoardInfoResponseSchema = z.object({
  data: z.array(GeneralBoardInfoSchema),
});

export type GeneralBoardInfo = z.infer<typeof GeneralBoardInfoSchema>;

export const fetchBoards = async (): Promise<GeneralBoardInfo[]> => {
  const response = await api.get('/boards');
  validateResponse(response);
  const result = GeneralBoardInfoResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}