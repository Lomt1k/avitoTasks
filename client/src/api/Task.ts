import { z } from "zod";
import { AssigneeSchema } from "./User";
import { api, validateResponse } from "./api";

export enum TaskStatus {
  Backlog = 'Backlog',
  InProgress = 'InProgress',
  Done = 'Done',
};

export enum TaskPriority {
  Low = 'Low',
  Medium = 'Medium',
  High = 'High',
}

export const TaskSchema = z.object({
  assignee: AssigneeSchema,
  boardId: z.number().optional(),
  boardName: z.string().optional(),
  description: z.string(),
  id: z.number(),
  priority: z.enum(Object.values(TaskPriority) as [string, ...string[]]),
  status: z.enum(Object.values(TaskStatus) as [string, ...string[]]),
  title: z.string(),
});

export const TasksResponseSchema = z.object({
  data: z.array(TaskSchema),
});

export const CreateTaskSchema = z.object({
  assigneeId: z.number(),
  boardId: z.number(),
  title: z.string()
    .nonempty('Необходимо указать название')
    .max(100, 'Должно быть не более 100 символов'),
  description: z.string()
    .nonempty('Необходимо указать описание')
    .max(500, 'Должно быть не более 500 символов'),
  priority: z.enum(Object.values(TaskPriority) as [string, ...string[]]),
});

export const UpdateTaskSchema = z.object({
  assigneeId: z.number(),
  title: z.string()
    .nonempty('Необходимо указать название')
    .max(100, 'Должно быть не более 100 символов'),
    description: z.string()
    .nonempty('Необходимо указать описание')
    .max(500, 'Должно быть не более 500 символов'),
  priority: z.enum(Object.values(TaskPriority) as [string, ...string[]]),
  status: z.enum(Object.values(TaskStatus) as [string, ...string[]]),
});

export type Task = z.infer<typeof TaskSchema>;

export type CreateTaskData = z.infer<typeof CreateTaskSchema>;

export type UpdateTaskData = z.infer<typeof UpdateTaskSchema>;

export const fetchTasksFromBoard = async (boardId: number): Promise<Task[]> => {
  const response = await api.get(`/boards/${boardId}`);
  validateResponse(response);
  const result = TasksResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}

export const fetchAllTasks = async (): Promise<Task[]> => {
  const response = await api.get(`/tasks`);
  validateResponse(response);
  const result = TasksResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}

export const fetchCreateTask = async (data: CreateTaskData): Promise<void> => {
  const response = await api.post('/tasks/create', data);
  validateResponse(response);
}

export const fetchUpdateTask = async (taskId: number, data: UpdateTaskData): Promise<void> => {
  const response = await api.put(`/tasks/update/${taskId}`, data);
  validateResponse(response);
}