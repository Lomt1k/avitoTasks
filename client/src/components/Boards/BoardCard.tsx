import { FC } from 'react';
import { GeneralBoardInfo } from '../../api/Boards';
import { ButtonLink } from '../ui';
import './BoardCard.scss';

type BoardCardProps = {
  board: GeneralBoardInfo;
}

const BoardCard: FC<BoardCardProps> = ({ board }) => {
  return (
    <div className="board-card">
      <div className="board-card__info">
        <div className="board-card__info-top">
          <span className="board-card__title">{board.name}</span>
          <span className="board-card__task-counter">{board.taskCount}</span>
        </div>
        <span>{board.description}</span>
      </div>
      <ButtonLink to={`/board/${board.id}`}>Перейти к доске</ButtonLink>
    </div>
  )
}

export { BoardCard }