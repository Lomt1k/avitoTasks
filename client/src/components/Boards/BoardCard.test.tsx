import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { BoardCard } from './BoardCard';
import { Board } from '../../api/Board';
import { BrowserRouter } from 'react-router';

const board: Board = {
  id: 1,
  name: 'Board',
  description: 'Description',
  taskCount: 5,
};

describe('BoardCard', () => {
  it('Рендерится корректно', () => {
    const { getByText } = render(
      <BrowserRouter>
        <BoardCard board={board} />
      </BrowserRouter>
    );

    expect(getByText('Board')).toBeInTheDocument();
    expect(getByText('Description')).toBeInTheDocument();
    expect(getByText('5')).toBeInTheDocument();
    expect(getByText('Перейти к доске')).toBeInTheDocument();
  });

  it('Снапшот', () => {
    const { container } = render(
      <BrowserRouter>
        <BoardCard board={board} />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
