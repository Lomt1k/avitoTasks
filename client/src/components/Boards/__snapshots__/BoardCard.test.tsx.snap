// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BoardCard > Снапшот 1`] = `
<div>
  <div
    class="board-card"
  >
    <div
      class="board-card__info"
    >
      <div
        class="board-card__info-top"
      >
        <span
          class="board-card__title"
        >
          Board
        </span>
        <span
          class="board-card__task-counter"
        >
          5
        </span>
      </div>
      <span>
        Description
      </span>
    </div>
    <a
      class="button"
      data-discover="true"
      href="/board/1"
    >
      Перейти к доске
    </a>
  </div>
</div>
`;

exports[`BoardCard > Создает снапшот 1`] = `
<div>
  <div
    class="board-card"
  >
    <div
      class="board-card__info"
    >
      <div
        class="board-card__info-top"
      >
        <span
          class="board-card__title"
        >
          Board
        </span>
        <span
          class="board-card__task-counter"
        >
          5
        </span>
      </div>
      <span>
        Description
      </span>
    </div>
    <a
      class="button"
      data-discover="true"
      href="/board/1"
    >
      Перейти к доске
    </a>
  </div>
</div>
`;
